addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,589 @@
+			return nil, err
+		}
+		r.Log.Info("cloned tanzuclusterbootstrap", "tanzuclusterbootstrap", tanzuClusterBootstrap)
+		return tanzuClusterBootstrap, nil
+	}
+
+	// TODO upgrade needs patch (update versions of all packages, merge configs, add additional packages and remove packages that don't exist anymore)
+	if tkrName != tanzuClusterBootstrap.Status.ResolvedTKR {
+		log.Info("TODO handle upgrade")
+		return nil, nil
+	}
+	return nil, errors.New("should not happen")
+}
+
+// createOrPatchPackageInstallSecret creates or patches or the secret used for PackageInstall in a cluster
+func (r *tanzuClusterBootstrapReconciler) createOrPatchPackageInstallSecret(cluster *clusterapiv1beta1.Cluster,
+	pkg *runtanzuv1alpha1.TanzuClusterBootstrapPackage, clusterClient client.Client, namespace string, log logr.Logger) (*corev1.Secret, error) {
+
+	secret := &corev1.Secret{}
+
+	if pkg.ValuesFrom.SecretRef != "" {
+		key := client.ObjectKey{Namespace: namespace, Name: pkg.ValuesFrom.SecretRef}
+		if err := r.Get(r.context, key, secret); err != nil {
+			log.Error(err, "unable to fetch secret", "objectkey", key)
+			return nil, err
+		}
+	}
+
+	if pkg.ValuesFrom.ProviderRef != nil {
+		gvr, err := r.getGVR(schema.GroupKind{Group: *pkg.ValuesFrom.ProviderRef.APIGroup, Kind: pkg.ValuesFrom.ProviderRef.Kind})
+		if err != nil {
+			log.Error(err, "failed to getGVR")
+			return nil, err
+		}
+		provider, err := r.dynamicClient.Resource(*gvr).Namespace(namespace).Get(r.context, pkg.ValuesFrom.ProviderRef.Name, metav1.GetOptions{}, "status")
+
+		if err != nil {
+			log.Error(err, "unable to fetch provider", "provider", pkg.ValuesFrom.ProviderRef, "gvr", gvr)
+			return nil, err
+		}
+		secretName, found, err := unstructured.NestedString(provider.UnstructuredContent(), "status", "secretRef")
+		if err != nil {
+			log.Error(err, "unable to fetch secretRef in provider", "provider", provider)
+			return nil, err
+		}
+		if !found {
+			log.Info("provider status does not have secretRef", "provider", provider)
+			return nil, nil
+		}
+		key := client.ObjectKey{Namespace: namespace, Name: secretName}
+		if err := r.Get(r.context, key, secret); err != nil {
+			log.Error(err, "unable to fetch secret", "objectkey", key)
+			return nil, err
+		}
+	}
+
+	dataValuesSecret := &corev1.Secret{}
+	dataValuesSecret.Name = fmt.Sprintf("%s-%s-data-values", cluster.Name, packageShortName(pkg.RefName))
+	dataValuesSecret.Namespace = constants.TKGSystemNS
+	dataValuesSecret.Type = corev1.SecretTypeOpaque
+
+	dataValuesSecretMutateFn := func() error {
+		dataValuesSecret.Data = map[string][]byte{}
+		for k, v := range secret.Data {
+			dataValuesSecret.Data[k] = v
+		}
+		return nil
+	}
+
+	_, err := controllerutil.CreateOrPatch(r.context, clusterClient, dataValuesSecret, dataValuesSecretMutateFn)
+	if err != nil {
+		log.Error(err, "Error creating or patching addon data values secret")
+		return nil, err
+	}
+	return dataValuesSecret, nil
+}
+
+//nolint:funlen
+//gocyclo:ignore
```

Comment:
I refactored to the maximum and gocyclo still complains that complexity 18 > 15. I am going to keep the lint comment and proposed a better complexity checker https://github.com/vmware-tanzu/tanzu-framework/issues/1563

Reply:
I know, gocyclo is pretty dumb, but if it's complaining, then the code is definitely not simple enough :)

From a closer look at the method, I can see it could definitely be simpler.

`secrets` and `providers` are actually not two completely different things. In fact, their purpose is exactly the same. If we could capture this purpose in an interface type, then we could collect and return one list of objects, not two. 

Sure, `*Secret` and `*Unstructured` don't directly do what we need. But we can wrap each in a struct implementing exactly our intended interface. In this particular case, `EnsureOwnerRef(*metav1.OwnerReference)` (I didn't find anything else). Our interface would look something like:

```go
type configProvider interface {
  EnsureOwnerRef(*metav1.OwnerReference) error
}
```

This would allow us to avoid creating an inline func (`aggregateSecretsProviders()` with 6 args!) mutating the two slices. Instead, `r.updateValues()` would return only two values: `(configProvider, error)`. This dramatically simplifies collecting and returning these to something like:

```go
configProvider, err := r.updateValues(/* args */)
if err != nil {
  return err
}
configProviders = append(configProviders, configProvider)
```

Also, since `CNI`, `CPI`, `CSI`, `Kapp` fields are of the same type, we can simply use a for loop over them:

```go
bootstrapPackages := append([]*runtanzuv1alpha3.TanzuClusterBootstrapPackage{
  bootstrap.Spec.CNI,
  bootstrap.Spec.CPI,
  bootstrap.Spec.CSI,
  bootstrap.Spec.Kapp,
}, bootstrap.Spec.AdditionalPackages...)

for _, bootstrapPackage := range bootstrapPackages {
  if bootstrapPackage == nil {
    continue
  }
  configProvider, err := r.updateValues(/* args */)
  if err != nil {
    return err
  }
  configProviders = append(configProviders, configProvider)
}
```

Reply:
List of `runtanzuv1alpha3.TanzuClusterBootstrapPackage` is a good suggestion, updating(guess needed a different perspective :). I am not sure about configProvider interface though.

Reply:
`configProvider` or another name () - doesn't matter. Maybe even `client.Object` could possibly fit here. The idea is: `createdSecrets` and `createdProviders` are only separate because they have a different element types. 

Their purpose in life is about the same thing though. If we could make that purpose into an interface type, we would further simplify the code by reducing "accidental" complexity.

Reply:
Had a chat with Ivan and this is being tracked in https://github.com/vmware-tanzu/tanzu-framework/issues/1610